Cache/Hashing.lean

```diff
@@ -0,0 +1,53 @@
+import Cache.IOUtils
+
+namespace Cache.Hashing
+
+open System IO
+
+/-- We store the root hash as a reader and cache the hash of each file for faster lookup -/
+abbrev HashM := ReaderT UInt64 $ StateT IO.HashMap IO
+
+/-- Gets the file paths to Mathlib files imported on a Lean source -/
```

Comment:
We also need to cache oleans for aesop/std4/quote4/etc.

Reply:
@semorrison was there some reason you asked me to cache Mathlib-specific build files?

Reply:
I just did some testing and it worked fine without those

Reply:
I'm not doubting that mathlib4 still builds.  But std4 takes almost as long to build as mathlib4 so we *need* to cache it.

Another important benchmark is: how long does it take to build a fresh checkout?
```
git clone https://github.com/leanprover-community/mathlib4
cd mathlib4
lake exe cache get
lake build
```

Reply:
Oh, that's what I meant by "it worked": it builds in less than 5 seconds:
```text
~/dev/lean/mathlib4 λ lake clean
~/dev/lean/mathlib4 λ rm -rf .cache
~/dev/lean/mathlib4 λ lake exe cache get
info: Building Cache.IOUtils
info: Compiling Cache.IOUtils
info: Building Cache.Hashing
info: Compiling Cache.Hashing
info: Building Cache.Requests
info: Building Cache.Main
info: Compiling Cache.Requests
info: Compiling Cache.Main
info: Linking cache
Downloading list of hosted files
Downloading 489 file(s)
DL% UL%  Dled  Uled  Xfers  Live   Qd Total     Current  Left    Speed
100 --  80.8M     0   489     0     0  0:00:04  0:00:08 --:--:-- 18.7M      
Decompressing cache
~/dev/lean/mathlib4 λ time lake build
Compiling Mathlib.Data.Array.Defs
Building scripts.runLinter
Building Mathlib
Compiling scripts.runLinter
Linking runLinter

real	0m2,515s
user	0m4,498s
sys	0m2,481s
```

Reply:
Another gotcha: there's nothing related to `std4` or any other lib in the `build` directory after `lake build`

Reply:
I see now that I was still keeping build files in `lake-packages`. I will work on caching those as well